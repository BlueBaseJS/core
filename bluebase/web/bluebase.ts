import { BootOptions } from '@bluebase/core';
import commonBootOptions from '../common/bluebase';
import deepmerge from 'deepmerge';

/**
 * Add your platform specific configs here. 
 * We keep all the universal (cross platform) configs in 
 * the common folder, and extend them here.
 */
const bootOptions: Partial<BootOptions> = {

	// config: {

	// 	wallpaper: {
	// 		backgroundColor: 'white',
	// 		resizeMode: 'cover',
	// 		source: require('./../../assets/web/wallpaper.png'),
	// 	},
	// }
};

export default deepmerge(commonBootOptions, bootOptions);
