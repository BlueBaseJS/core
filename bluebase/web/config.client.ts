export default function (input: any) {
	return input;
}