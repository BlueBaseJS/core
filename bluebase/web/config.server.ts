function config(input: any) {
	return input;
}

export default config;
