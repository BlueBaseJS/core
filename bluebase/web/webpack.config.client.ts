/* eslint-disable max-len */
import { DefinePlugin } from 'webpack';
import path from 'path';

export default function(config: any) {
	config.resolve.alias = {
		...config.resolve.alias,

		'@bluebase/core': path.resolve(__dirname, '../../src'),

		'react-native$': 'react-native-web',
		// tslint:disable-next-line: object-literal-sort-keys
		'@react-native-community/netinfo': 'react-native-web/dist/exports/NetInfo',
		// Add polyfills for modules that react-native-web doesn't support
		// Depends on expo-asset
		'react-native/Libraries/Image/AssetSourceResolver$': 'expo-asset/build/AssetSourceResolver',
		'react-native/Libraries/Image/assetPathUtils$': 'expo-asset/build/Image/assetPathUtils',
		'react-native/Libraries/Image/resolveAssetSource$': 'expo-asset/build/resolveAssetSource',
		// Alias internal react-native modules to react-native-web
		'react-native/Libraries/Components/View/ViewStylePropTypes$':
			'react-native-web/dist/exports/View/ViewStylePropTypes',
		'react-native/Libraries/EventEmitter/RCTDeviceEventEmitter$':
			'react-native-web/dist/vendor/react-native/NativeEventEmitter/RCTDeviceEventEmitter',
		'react-native/Libraries/vendor/emitter/EventEmitter$':
			'react-native-web/dist/vendor/react-native/emitter/EventEmitter',
		'react-native/Libraries/vendor/emitter/EventSubscriptionVendor$':
			'react-native-web/dist/vendor/react-native/emitter/EventSubscriptionVendor',
		'react-native/Libraries/EventEmitter/NativeEventEmitter$':
			'react-native-web/dist/vendor/react-native/NativeEventEmitter',
	};

	config.module.rules.push({
		test: /\.(ts|tsx)$/,

		include: [
			path.resolve(__dirname, '../../../src'),
			path.resolve(__dirname, '../../../bluebase'),
		],
		use: [
			{
				loader: require.resolve('ts-loader'),
			},
			// require.resolve('react-docgen-typescript-loader'),
		],
	});

	config.module.rules.push({
		// tslint:disable-next-line: max-line-length
		exclude: /node_modules\/(?!(rn-placeholder|react-native-elements|react-native-vector-icons|expo-image-picker|react-native-animatable|react-native-svg|react-native-reanimated|expo-face-detector|react-native-gesture-handler|expo-linear-gradient|expo-payments-stripe|expo-gl|expo-av|expo-camera|expo|expo-barcode-scanner|expo-ads-admob|expo-blur|@expo\/vector-icons|@unimodules\/react-native-adapter)\/).*/,
		loader: 'babel-loader',
		test: /\.js$/,
	});

	// config = withUnimodules(config, {
	// 	config: path.resolve(__dirname, '../../../build/storybook-native/app.json'),
	// });

	config.resolve.extensions = [
		'.web.ts',
		'.web.tsx',
		'.ts',
		'.tsx',
		'.web.js',
		'.web.jsx',
		'.js',
		'.jsx',
		'.json',
	];

	config.plugins.push(new DefinePlugin({ __DEV__: config.mode !== 'production' }));

	return config;
}
