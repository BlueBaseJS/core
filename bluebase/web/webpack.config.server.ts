import path from 'path';

export default function (input: any) {

	input.resolve.alias['@bluebase/core'] = path.resolve(__dirname, '../../src');

	return input;
}