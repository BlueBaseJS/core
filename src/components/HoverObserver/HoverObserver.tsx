import { HoverObserverProps } from '@bluebase/components';

import { renderChildrenWithProps } from '../../utils';

export interface HoverObserverState {
	readonly isHovering: boolean;
}

export const HoverObserver = (props: HoverObserverProps) => {
	return renderChildrenWithProps(props.children, { isHovering: false });
};

HoverObserver.displayName = 'HoverObserver';
