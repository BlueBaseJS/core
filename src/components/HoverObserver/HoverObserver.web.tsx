import React, { SyntheticEvent } from 'react';
import { HoverObserverProps } from '@bluebase/components';
import { renderChildrenWithProps } from '../../utils';

export interface HoverObserverState {
	readonly isHovering: boolean;
}

const Noop = () => { return; };

/**
 * # ðŸ›¸ HoverObserver
 *
 * A React component that notifies its children of hover interactions.
 *
 * Initial code taken from: https://github.com/ethanselzer/react-hover-observer.
 *
 * ## Usage
 * ```jsx
 * <HoverObserver>
 *  {({ isHovering }) => (
 * 		 <YourChildComponent isActive={isHovering} />
 *  )}
 * </HoverObserver>
 * ```
 */
export class HoverObserver extends React.PureComponent<HoverObserverProps, HoverObserverState> {

	public static defaultProps: Partial<HoverObserverProps> = {
		hoverDelayInMs: 0,
		hoverOffDelayInMs: 0,
		onHoverChanged: Noop,
		onMouseEnter: ({ setIsHovering }) => setIsHovering(),
		onMouseLeave: ({ unsetIsHovering }) => unsetIsHovering(),
		onMouseOut: Noop,
		onMouseOver: Noop,
	};

	private timerIds: number[] = [];

	constructor(props: HoverObserverProps) {
		super(props);

		this.state = {
			isHovering: false
		};

		this.onMouseEnter = this.onMouseEnter.bind(this);
		this.onMouseLeave = this.onMouseLeave.bind(this);
		this.onMouseOver = this.onMouseOver.bind(this);
		this.onMouseOut = this.onMouseOut.bind(this);
		this.setIsHovering = this.setIsHovering.bind(this);
		this.unsetIsHovering = this.unsetIsHovering.bind(this);
		this.componentWillUnmount = this.componentWillUnmount.bind(this);

		this.timerIds = [];
	}

	onMouseEnter(e: SyntheticEvent) {
		this.props.onMouseEnter!({
			e,
			setIsHovering: this.setIsHovering,
			unsetIsHovering: this.unsetIsHovering
		});
	}

	onMouseLeave(e: SyntheticEvent) {
		this.props.onMouseLeave!({
			e,
			setIsHovering: this.setIsHovering,
			unsetIsHovering: this.unsetIsHovering
		});
	}

	onMouseOver(e: SyntheticEvent) {
		this.props.onMouseOver!({
			e,
			setIsHovering: this.setIsHovering,
			unsetIsHovering: this.unsetIsHovering
		});
	}

	onMouseOut(e: SyntheticEvent) {
		this.props.onMouseOut!({
			e,
			setIsHovering: this.setIsHovering,
			unsetIsHovering: this.unsetIsHovering
		});
	}

	componentWillUnmount() {
		this.clearTimers();
	}

	setIsHovering() {
		this.clearTimers();

		const hoverScheduleId = setTimeout(() => {
			const newState = { isHovering: true };
			this.setState(newState, () => {
				this.props.onHoverChanged!(newState);
			});
		}, this.props.hoverDelayInMs) as any;

		this.timerIds.push(hoverScheduleId);
	}

	unsetIsHovering() {
		this.clearTimers();

		const hoverOffScheduleId = setTimeout(() => {
			const newState = { isHovering: false };
			this.setState(newState, () => {
				this.props.onHoverChanged!(newState);
			});
		}, this.props.hoverOffDelayInMs) as any;

		this.timerIds.push(hoverOffScheduleId);
	}

	clearTimers() {
		const ids = this.timerIds;
		while (ids.length) {
			const id = ids.pop() as number;
			clearTimeout(id);
		}
	}

	render() {
		const { children } = this.props;

		return (
			<div {...{
				onMouseEnter: this.onMouseEnter,
				onMouseLeave: this.onMouseLeave,
				onMouseOut: this.onMouseOut,
				onMouseOver: this.onMouseOver,
			}}>
				{renderChildrenWithProps(children, this.state)}
			</div>
		);
	}
}