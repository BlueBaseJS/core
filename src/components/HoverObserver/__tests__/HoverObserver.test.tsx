import { HoverObserver } from '../HoverObserver';
import { HoverObserver as HoverObserverWeb } from '../HoverObserver.web';
import React from 'react';
import { Text } from 'react-native';
import { mount } from 'enzyme';

jest.useFakeTimers();

describe('HoverObserver', () => {

	it(`should render a HoverObserver with default props`, () => {

		const wrapper = mount(
			<HoverObserverWeb>
			{({ isHovering }) => <Text>{isHovering ? 'Hovering!' : 'Not Hovering!'}</Text>}
			</HoverObserverWeb>
		);

		// Check component
		expect(wrapper).toMatchSnapshot();
		expect(wrapper.find('HoverObserver Text').last().text()).toBe('Not Hovering!');

		// Change state: mouseEnter
		wrapper.find('HoverObserver div').simulate('mouseEnter');
		jest.advanceTimersByTime(1);
		wrapper.update();

		// Check component
		expect(wrapper).toMatchSnapshot();
		expect(wrapper.find('HoverObserver Text').last().text()).toBe('Hovering!');

		// Change state: mouseLeave
		wrapper.find('HoverObserver div').simulate('mouseLeave');
		jest.advanceTimersByTime(1);
		wrapper.update();

		// Check component
		expect(wrapper).toMatchSnapshot();
		expect(wrapper.find('HoverObserver Text').last().text()).toBe('Not Hovering!');

		// Change state: mouseOver
		expect((wrapper as any).find('HoverObserver').instance().onMouseOver()).toBeUndefined();
		expect((wrapper as any).find('HoverObserver').instance().onMouseOut()).toBeUndefined();

		wrapper.unmount();
	});

	it(`should always set isHovering to false on android`, () => {

		const wrapper = mount(
			<HoverObserver>
			{({ isHovering }: any) => <Text>{isHovering ? 'Hovering!' : 'Not Hovering!'}</Text>}
			</HoverObserver>
		);

		// Check component
		expect(wrapper).toMatchSnapshot();
		expect(wrapper.find('HoverObserver Text').last().text()).toBe('Not Hovering!');

		// Change state: mouseEnter
		wrapper.find('HoverObserver Text').first().simulate('mouseEnter');
		jest.advanceTimersByTime(1);
		wrapper.update();

		// Check component
		expect(wrapper).toMatchSnapshot();
		expect(wrapper.find('HoverObserver Text').last().text()).toBe('Not Hovering!');
	});

});
