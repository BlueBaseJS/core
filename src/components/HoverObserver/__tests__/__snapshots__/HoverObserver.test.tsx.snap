// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HoverObserver should always set isHovering to false on android 1`] = `
<HoverObserver>
  <Component>
    <Text>
      Not Hovering!
    </Text>
  </Component>
</HoverObserver>
`;

exports[`HoverObserver should always set isHovering to false on android 2`] = `
<HoverObserver>
  <Component>
    <Text>
      Not Hovering!
    </Text>
  </Component>
</HoverObserver>
`;

exports[`HoverObserver should render a HoverObserver with default props 1`] = `
<HoverObserver
  hoverDelayInMs={0}
  hoverOffDelayInMs={0}
  onHoverChanged={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <div
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    <Component>
      <Text>
        Not Hovering!
      </Text>
    </Component>
  </div>
</HoverObserver>
`;

exports[`HoverObserver should render a HoverObserver with default props 2`] = `
<HoverObserver
  hoverDelayInMs={0}
  hoverOffDelayInMs={0}
  onHoverChanged={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <div
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    <Component>
      <Text>
        Hovering!
      </Text>
    </Component>
  </div>
</HoverObserver>
`;

exports[`HoverObserver should render a HoverObserver with default props 3`] = `
<HoverObserver
  hoverDelayInMs={0}
  hoverOffDelayInMs={0}
  onHoverChanged={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <div
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    <Component>
      <Text>
        Not Hovering!
      </Text>
    </Component>
  </div>
</HoverObserver>
`;
